### PR TITLE
Fix Iceberg v3 structured type VARCHAR mismatch

### LINKED ISSUE #1721: [Bug] Iceberg v3 incremental microbatch fails on batch 2+ due to VARCHAR(134217728) vs VARCHAR(16777216) mismatch in structured types
### Is this a new bug?

- [x] I believe this is a new bug
- [x] I have searched the existing issues, and I could not find an existing issue for this bug

### Which packages are affected?

- [x] dbt-snowflake

### Current Behavior

When using Iceberg v3 tables (Snowflake-managed, `catalog_type: built_in`) with `incremental_strategy: microbatch` and `on_schema_change: fail`, batch 1 succeeds (CTAS creates the table) but batch 2+ fails with a schema mismatch error on columns containing structured types (e.g., `OBJECT(...)`).

**Root cause:** Snowflake Iceberg v3 tables store ALL `STRING`/`VARCHAR` columns as `VARCHAR(134217728)` (Iceberg's max), not `VARCHAR(16777216)` (Snowflake's native default). This can be confirmed with a simple test:

```sql
CREATE OR REPLACE ICEBERG TABLE test (
    val_one STRING,
    val_two OBJECT(test STRING)
)
EXTERNAL_VOLUME = 'MY_VOLUME'
CATALOG = 'SNOWFLAKE';

DESC TABLE test;
-- VAL_ONE: VARCHAR(134217728)
-- VAL_TWO: OBJECT(test VARCHAR(134217728))
```

When `process_schema_changes` compares the existing table schema (from `DESC TABLE`) against the new batch's column definitions, it sees `VARCHAR(134217728)` in the table vs `VARCHAR(16777216)` from the SQL/contract, treats it as a type change, and raises an error.

**Note:** The fix in dbt-snowflake 1.10.6/1.11.1 (for #1258) may address top-level VARCHAR columns, but structured type fields inside `OBJECT(...)` are not covered — the comparison operates on the full type string including nested field definitions.

### Expected Behavior

Batch 2+ should succeed. `VARCHAR(134217728)` and `VARCHAR(16777216)` (or `STRING`) should be treated as compatible types during schema comparison, including inside structured `OBJECT(...)` type definitions.

### Steps To Reproduce

1. Create a `catalogs.yml` with a Snowflake-managed Iceberg catalog:
```yaml
catalogs:
  - name: my_catalog
    active_write_integration: write
    write_integrations:
      - name: write
        external_volume: MY_VOLUME
        table_format: iceberg
        catalog_type: built_in
```

2. Create an incremental microbatch model with `on_schema_change: fail`, contract enforcement, and a structured `OBJECT(...)` column:
```yaml
config:
  materialized: incremental
  incremental_strategy: microbatch
  batch_size: day
  on_schema_change: fail
  contract: {enforced: true}
  catalog_name: my_catalog
columns:
  - name: my_object_col
    data_type: >
      object(
        name string,
        amount number(38,0)
      )
```

3. Run `dbt build`. Batch 1 creates the table via CTAS — succeeds.

4. Batch 2 attempts incremental merge. `DESC TABLE` shows the object fields as `VARCHAR(134217728)`, but the new batch produces `VARCHAR(16777216)`. `process_schema_changes` detects a mismatch and fails.

### Relevant log output

```
Compilation Error in model my_model

  The source and target schemas on this incremental model are out of sync!

  Additional troubleshooting context:
     Source columns not in target: []
     Target columns not in source: []
     New column types: [
       {'column_name': 'MY_OBJECT_COL',
        'new_type': 'OBJECT(name VARCHAR(16777216), amount NUMBER(38,0))'}
     ]
```

The table (from `DESC TABLE`) actually has:
```
OBJECT(name VARCHAR(134217728), amount NUMBER(38,0))
```

### Environment

```markdown
- dbt: dbt Cloud CLI
- dbt-snowflake: embedded in dbt Cloud
- Snowflake: Iceberg v3 (recently enabled), Snowflake-managed catalog (BUILT_IN)
```

### Additional Context

- This is specifically with **Iceberg v3 spec** which Snowflake recently started supporting
- Related to #1258 which addressed VARCHAR length mismatches for top-level columns with collation, but did not cover structured type fields
- Related to #1516 which covers `ALTER ICEBERG TABLE` vs `ALTER TABLE` command differences
- Workaround: use `VARCHAR(134217728)` explicitly in the YAML contract's object type definition, but this is fragile and couples the contract to Iceberg-specific implementation details

### PR DESCRIPTION
**Pull request by jecolvin**
## Summary
- Normalizes `VARCHAR(134217728)` (Iceberg max) to `VARCHAR(16777216)` (Snowflake native max) inside structured type definitions (OBJECT, ARRAY, MAP) when parsing columns in `SnowflakeColumn.from_description()`
- Iceberg v3 tables store all VARCHAR fields as `VARCHAR(134217728)`, while contract-enforced models define them as `VARCHAR(16777216)`, causing false schema mismatch errors during incremental runs
- Top-level VARCHAR columns were already handled by existing `can_expand_to()` logic; this fix covers structured types where the raw type string is compared directly

Closes #1721

## Test plan
- [x] Added 11 unit tests in `TestSnowflakeColumnIcebergVarcharNormalization` covering OBJECT, ARRAY, MAP normalization, edge cases, case-insensitivity, and schema comparison equivalence
- [x] Verified existing `TestSnowflakeColumnStructuredTypes` and `TestSnowflakeColumnCollation` tests still pass with no regressions
- [ ] Functional testing against Snowflake Iceberg v3 environment (requires Snowflake connection)

### CODE REVIEW

File: dbt-snowflake/tests/unit/test_snowflake_adapter.py [hunk 1]
```diff
@@ -1009,6 +1009,73 @@ def test_map_from_description(self):
         assert col.is_array() is False
 
 
+class TestSnowflakeColumnIcebergVarcharNormalization(unittest.TestCase):
+    """
+    Iceberg v3 tables store VARCHAR as VARCHAR(134217728) instead of Snowflake's
+    native VARCHAR(16777216). This causes false schema mismatch errors in structured
+    types during incremental runs. Verify normalization in from_description().
+    See: https://github.com/dbt-labs/dbt-adapters/issues/1721
+    """
```
[the remark this answers is not in the document]
Comment (jecolvin) [reply dbt-snowflake/tests/unit/test_snowflake_adapter.py:1018]: @copilot apply changes based on [this feedback](https://github.com/dbt-labs/dbt-adapters/pull/1729#discussion_r2947100351)

**Comment by jecolvin** [comment]
@colin-rogers-dbt Fixed the black failures!

**Comment by colin-k-rogers** [comment]
@jecolvin can you add a repro example to our existing iceberg functional tests?

**Comment by jecolvin** [comment]
@colin-rogers-dbt Added!

File: dbt-snowflake/tests/unit/test_snowflake_adapter.py [hunk 2]
```diff
@@ -1009,6 +1009,67 @@ def test_map_from_description(self):
...
```
[the remark this answers is not in the document]
Comment (jecolvin) [reply dbt-snowflake/tests/unit/test_snowflake_adapter.py:1018]: Thera are indeed 11 tests total — the 10 in `TestSnowflakeColumnIcebergVarcharNormalization` plus 1 in
  `TestSnowflakeColumnStructuredTypes`

File: dbt-snowflake/src/dbt/adapters/snowflake/column.py [hunk 3]
```diff
@@ -94,6 +98,15 @@ def from_description(cls, name: str, raw_data_type: str) -> "SnowflakeColumn":
             collation=collation,
         )
 
+    @staticmethod
+    def _normalize_iceberg_varchar(raw_data_type: str) -> str:
+        """Normalize Iceberg VARCHAR(134217728) to Snowflake VARCHAR(16777216) in structured types."""
+        return re.sub(
+            r"(?i)\bVARCHAR\(" + str(ICEBERG_MAX_VARCHAR_LENGTH) + r"\)",
+            f"VARCHAR({SNOWFLAKE_MAX_VARCHAR_LENGTH})",
+            raw_data_type,
```
[the remark this answers is not in the document]
Comment (jecolvin) [reply dbt-snowflake/src/dbt/adapters/snowflake/column.py:111]: Done — updated the docstring to mention all composite types (OBJECT, ARRAY, MAP, VECTOR).

File: dbt-snowflake/src/dbt/adapters/snowflake/column.py [hunk 4]
```diff
@@ -70,7 +73,8 @@ def from_description(cls, name: str, raw_data_type: str) -> "SnowflakeColumn":
 
         # We want to pass through numeric parsing for composite types
```
[the remark this answers is not in the document]
Comment (jecolvin) [reply dbt-snowflake/src/dbt/adapters/snowflake/column.py]: Done — replaced with a more accurate comment describing the opaque passthrough + Iceberg VARCHAR normalization.

File: dbt-snowflake/src/dbt/adapters/snowflake/column.py [hunk 5]
```diff
@@ -94,6 +98,15 @@ def from_description(cls, name: str, raw_data_type: str) -> "SnowflakeColumn":
...
         )
 
+    @staticmethod
+    def _normalize_iceberg_varchar(raw_data_type: str) -> str:
+        """Normalize Iceberg VARCHAR(134217728) to Snowflake VARCHAR(16777216) in structured types."""
+        return re.sub(
+            r"(?i)\bVARCHAR\(" + str(ICEBERG_MAX_VARCHAR_LENGTH) + r"\)",
+            f"VARCHAR({SNOWFLAKE_MAX_VARCHAR_LENGTH})",
+            raw_data_type,
+        )
```
[the remark this answers is not in the document]
Comment (jecolvin) [reply dbt-snowflake/src/dbt/adapters/snowflake/column.py:112]: Done — precompiled the pattern as `ICEBERG_VARCHAR_PATTERN` at module scope.

**Comment by jecolvin** [comment]
@colin-rogers-dbt I've addressed all the copilot feedback now as well
